adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/MarkDuplicates.scala

```diff
@@ -17,153 +17,419 @@
  */
 package org.bdgenomics.adam.rdd.read
 
-import org.bdgenomics.utils.misc.Logging
 import org.apache.spark.rdd.RDD
-import org.bdgenomics.adam.instrumentation.Timers._
-import org.bdgenomics.adam.models.{
-  RecordGroupDictionary,
-  ReferencePosition
-}
+import org.apache.spark.sql._
+import org.apache.spark.sql.expressions.Window
+import org.apache.spark.sql.functions.{ countDistinct, first, row_number, sum, when }
+import org.bdgenomics.adam.models.RecordGroupDictionary
 import org.bdgenomics.adam.rdd.ADAMContext._
 import org.bdgenomics.adam.rdd.fragment.FragmentRDD
-import org.bdgenomics.formats.avro.{ AlignmentRecord, Fragment }
+import org.bdgenomics.adam.sql.{ AlignmentRecord => AlignmentRecordSchema, Fragment => FragmentSchema }
+import org.bdgenomics.formats.avro.{ AlignmentRecord, Fragment, Strand }
+import org.bdgenomics.utils.misc.Logging
+import htsjdk.samtools.{ Cigar, CigarElement, CigarOperator, TextCigarCodec }
+import scala.collection.JavaConversions._
 
 private[rdd] object MarkDuplicates extends Serializable with Logging {
 
-  private def markReadsInBucket(bucket: SingleReadBucket, primaryAreDups: Boolean, secondaryAreDups: Boolean) {
-    bucket.primaryMapped.foreach(read => {
-      read.setDuplicateRead(primaryAreDups)
-    })
-    bucket.secondaryMapped.foreach(read => {
-      read.setDuplicateRead(secondaryAreDups)
-    })
-    bucket.unmapped.foreach(read => {
-      read.setDuplicateRead(false)
-    })
+  /**
+   * Marks alignment records as PCR duplicates.
+   * This class marks duplicates all read pairs that have the same pair alignment locations,
+   * and all unpaired reads that map to the same sites. Only the highest scoring
+   * read/read pair is kept, where the score is the sum of all quality scores minus 33 in
+   * the read that are greater than 15.
+   * @param alignmentRecords GenomicRDD of alignment records
+   * @return RDD of alignment records with the "duplicateRead" field marked appropriately
+   */
+  def apply(alignmentRecords: AlignmentRecordRDD): RDD[AlignmentRecord] = {
```

Comment:
We should refactor the caller here so that we don't force the conversions between RDD and Dataset.  In other words, only perform the conversions if `alignmentRecords` has been realized as `RDDBoundAlignmentRecordRDD`.

Reply:
I see, so there should be another `apply` method which takes `Dataset[AlignmentRecord]` and this `apply` method's signature should be changed to take `RDDBoundAlignmentRecordRDD`?

Reply:
Or do you mean that this method should just be changed to take `Dataset[AlignmentRecord]`. In this case would it also make sense to refactor so that the other `apply` method also just takes `Dataset[Fragment]` and do a similar refactoring of the caller?

Reply:
Does this look right?  We would want the caller to be
```scala
abstract class AlignmentRecordRDD ... {
  def markDuplicates(): AlignmentRecordRDD = {
    replaceRdd(MarkDuplicates(this.rdd, this.recordGroups))
  }
}

case class DatasetBoundAlignmentRecordRDD ... {
  override def markDuplicates(): AlignmentRecordRDD = {
    replaceDataset(MarkDuplicates(this.dataset, this.recordGroups))
  }
}

abstract class FragmentRDD ... {
  def markDuplicates(): FragmentRDD = {
    replaceRdd(MarkDuplicates(this.rdd, this.recordGroups))
  }
}

case class DatasetBoundFragmentRDD ... {
  override def markDuplicates(): FragmentRDD = {
    replaceDataset(MarkDuplicates(this.dataset, this.recordGroups))
  }
}
```

so then the `apply` methods might be
```scala
import org.bdgenomics.formats.avro.{
  AlignmentRecord,
  Fragment
}
import org.bdgenomics.adam.sql.{
  AlignmentRecord => AlignmentRecordProduct,
  Fragment => FragmentProduct
}

object MarkDuplicates {
  def apply(RDD[AlignmentRecord], RecordGroupDictionary): RDD[AlignmentRecord] = { }
  def apply(RDD[Fragment], RecordGroupDictionary): RDD[Fragment] = { }
  def apply(Dataset[AlignmentRecordProduct], RecordGroupDictionary): Dataset[AlignmentRecordProduct] = { }
  def apply(Dataset[FragmentProduct], RecordGroupDictionary): Dataset[FragmentProduct] = { }
}
```

Reply:
Alright sounds good. Because of type erasure there will need to be a single `apply[T]` for `RDD` and for `Dataset` for fragments/alignment-records.

Reply:
Ah, good point.  Is `RecordGroupDictionary` actually required for the fragment cases?  If not, that could be the discriminator.

Reply:
I believe that it is.

Reply:
Would it be better to take `RDDBoundAlignmentRecordRDD` and `DatasetBoundAlignmentRecordRDD` so that the code for converting from `RDD` to `Dataset` that is already implemented in `RDDBoundAlignmentRecordRDD` is not duplicated in `MarkDuplicates`?